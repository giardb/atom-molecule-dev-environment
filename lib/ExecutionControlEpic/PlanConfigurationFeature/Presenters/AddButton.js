"use babel";
// @flow

"use babel";
// @flow

import React from "react";
import classNames from "classnames";

export default class AddButton extends React.Component<
  DefaultProps,
  Props,
  State,
> {
  state: State;
  props: Props;
  static defaultProps: DefaultProps;

  constructor(props: Props) {
    super(props);
  }

  render() {
    return (
      <button
        onClick={this.props.onClick}
        style={{ margin: "10px 5px", alignSelf: "flex-end" }}
        className={classNames("btn", "icon", "icon-plus", {
          "btn-success": this.props.success,
        })}
      >
        {this.props.children}
      </button>
    );
  }
}

AddButton.defaultProps = {
  success: false,
};

type DefaultProps = {
  success: false,
};

type Props = {
  onClick(): void,
  children?: string,
  success: boolean,
};

type State = {};
