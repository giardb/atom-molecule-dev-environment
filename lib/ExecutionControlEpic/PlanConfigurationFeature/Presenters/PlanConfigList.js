"use babel";
// @flow

import React from "react";
import PlanConfigPart from "./PlanConfigPart";
import type { ConfigSchemaPart } from "../Types/types.js.flow";
import { fromJS } from "immutable";
import { getDefault } from "../Model/PlanConfigManipulators";
import AddButton from "./AddButton";
import styled from "styled-components";
import { compose } from "recompose";
// $FlowFixMe
import { StateProxy } from "react-forms-state";

const Box = styled.div`
  display: flex;
  flex-direction: column;
  align-items: stretch;
  margin: 8px 0px 8px 0px;
  border-radius: 5px;
`;

const Label = styled.span.attrs({
  className: "text-color-highlight",
})`
  display: flex;
  font-size: 13px;
`;

const List = styled.ul`
  list-style: none;
  padding: 0px;
  margin: 8px 0px;
  display: flex;
  flex-direction: column;
  align-items: stretch;
`;

const Item = styled.li`
  display: flex;
  flex-direction: column;
  list-style: none;
  align-items: stretch;
`;

export function PlanConfigList({
  label,
  values = [],
  onChange,
  items,
}: {
  label: string,
  value: Array<mixed>,
  onChange: (v: Array<mixed>) => void,
  items: ConfigSchemaPart,
}) {
  return (
    <Box>
      <Label>{label}</Label>
      <List>
        {values.map((value, i) => (
          <Item key={i}>
            <PlanConfigPart
              // $FlowFixMe
              {...items}
              value={value}
              onChange={v =>
                onChange(fromJS(this.props.values).set(i, v).toJS())}
            />
          </Item>
        ))}
        <AddButton
          onClick={() => onChange([].concat(values).concat(getDefault(items)))}
        />
      </List>
    </Box>
  );
}

const enhance = compose(StateProxy());

export default enhance(PlanConfigList);
