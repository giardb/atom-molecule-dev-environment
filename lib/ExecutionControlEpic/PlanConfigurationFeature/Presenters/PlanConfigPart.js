"use babel";
// @flow

import React from "react";
import PlanConfigSelectInputField from "./PlanConfigSelectInputField";
import PlanConfigTextInputField from "./PlanConfigTextInputField";
import PlanConfigGroup from "./PlanConfigGroup";
import type {
  ArrayConfigSchema,
  ConditionalConfigSchema,
  ConfigSchemaPartBase,
  EnumConfigSchema,
  NumberConfigSchema,
  ObjectConfigSchema,
  StringConfigSchema,
  FileConfigSchema,
} from "../Types/types.js.flow";
import PlanConfigConditionalGroup from "./PlanConfigConditionalGroup";
import PlanConfigNumberInputField from "./PlanConfigNumberInputField";
import PlanConfigCheckboxInputField from "./PlanConfigCheckboxInputField";
import PlanConfigFileInputField from "../Containers/PlanConfigFileInputField";
import PlanConfigList from "./PlanConfigList";

export default function PlanConfigPart({
  type,
  value,
  onChange,
  label,
  items,
  expression,
  cases,
  placeholder,
  schemas,
  tester,
  ...props
}: ConfigSchemaPartBase & {
  onChange: (value: any) => void,
  // eslint-disable-next-line prettier/prettier
} & ( // eslint-disable-next-line prettier/prettier
    | (EnumConfigSchema & { value: any })
    | (FileConfigSchema & { value: any })
    | (ObjectConfigSchema & { value: { [key: string]: any } })
    | (StringConfigSchema & { value: string })
    | (NumberConfigSchema & { value: number })
    | (ConditionalConfigSchema & {
      value: { expressionValue: mixed, caseValue: mixed },
    })
    | (ArrayConfigSchema & { value: Array<any> }))) {
  switch (type) {
    case "enum":
      return (
        <PlanConfigSelectInputField
          options={props.enum}
          default={props.default}
          value={value}
          onChange={onChange}
        />
      );
    case "file":
      return (
        <PlanConfigFileInputField
          tester={tester}
          default={props.default}
          value={value}
          label={label}
          onChange={onChange}
        />
      );
    case "object":
      return (
        <PlanConfigGroup
          label={label}
          schemas={schemas}
          value={value}
          onChange={onChange}
        />
      );
    case "string":
      return (
        <PlanConfigTextInputField
          default={props.default}
          value={value}
          onChange={onChange}
          label={label}
          placeholder={placeholder}
        />
      );
    case "number":
      return (
        <PlanConfigNumberInputField
          default={props.default}
          value={value}
          onChange={onChange}
          label={label}
        />
      );
    case "conditional":
      return (
        <PlanConfigConditionalGroup
          default={props.default}
          expression={expression}
          cases={cases}
          value={value}
          onChange={onChange}
        />
      );
    case "array":
      return (
        <PlanConfigList
          values={value}
          default={props.default}
          onChange={onChange}
          items={items}
          label={label}
        />
      );
    case "boolean":
      return (
        <PlanConfigCheckboxInputField
          value={value}
          default={props.default}
          onChange={onChange}
          label={label}
        />
      );
    default:
      return null;
  }
}
