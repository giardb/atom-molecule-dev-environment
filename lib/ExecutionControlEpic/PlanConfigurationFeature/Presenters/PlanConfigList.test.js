import React from "react";
import { mount } from "enzyme";
import { PlanConfigList } from "./PlanConfigList";
import PlanConfigPart from "./PlanConfigPart";
import AddButton from "./AddButton";

describe("PlanConfigList", () => {
  let spy = jest.fn();
  let subject = mount(
    <PlanConfigList
      values={[
        { name: "NODE_PLAN", value: "web" },
        { name: "NODE_ENV", value: "dev" },
      ]}
      items={{
        type: "object",
        schemas: {
          name: {
            type: "string",
            default: "",
            placeholder: "name",
          },
          value: {
            type: "string",
            default: "",
            placeholder: "value",
          },
        },
      }}
      onChange={spy}
    />,
  );
  it("should display the values", () => {
    expect(subject.find(PlanConfigPart).length).toBe(6); // nested parts
    expect(subject.find(PlanConfigPart).at(0).prop("value")).toEqual({
      name: "NODE_PLAN",
      value: "web",
    });
    expect(subject.find(PlanConfigPart).at(0).prop("type")).toBe("object");
    expect(subject.find(PlanConfigPart).at(0).prop("schemas")).toEqual({
      name: {
        type: "string",
        default: "",
        placeholder: "name",
      },
      value: {
        type: "string",
        default: "",
        placeholder: "value",
      },
    });
  });

  it("should display an add button", () => {
    expect(subject.find(AddButton).length).toBe(1);
  });

  it("should add a value when button is clicked", () => {
    subject.find(AddButton).at("0").simulate("click", {});

    expect(spy).toBeCalledWith([
      { name: "NODE_PLAN", value: "web" },
      { name: "NODE_ENV", value: "dev" },
      { name: "", value: "" },
    ]);
  });
});
