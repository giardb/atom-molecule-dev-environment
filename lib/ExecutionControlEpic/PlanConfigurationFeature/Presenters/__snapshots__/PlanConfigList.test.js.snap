// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanConfigList should display the values 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div>
    <styled.span
      error={false}
    />
    <StateInjector
      __debug={false}
      watchPath={[Function]}
    >
      [Function]
    </StateInjector>
  </styled.div>,
  "nodes": Array [
    <styled.div>
      <styled.span
        error={false}
      />
      <StateInjector
        __debug={false}
        watchPath={[Function]}
      >
        [Function]
      </StateInjector>
    </styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <PlanConfigList
        items={
          Object {
            "schemas": Object {
              "name": Object {
                "default": "",
                "placeholder": "name",
                "type": "string",
              },
              "value": Object {
                "default": "",
                "placeholder": "value",
                "type": "string",
              },
            },
            "type": "object",
          }
        }
        onChange={[Function]}
        values={
          Array [
            Object {
              "name": "NODE_PLAN",
              "value": "web",
            },
            Object {
              "name": "NODE_ENV",
              "value": "dev",
            },
          ]
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "items": Object {
            "schemas": Object {
              "name": Object {
                "default": "",
                "placeholder": "name",
                "type": "string",
              },
              "value": Object {
                "default": "",
                "placeholder": "value",
                "type": "string",
              },
            },
            "type": "object",
          },
          "onChange": [Function],
          "values": Array [
            Object {
              "name": "NODE_PLAN",
              "value": "web",
            },
            Object {
              "name": "NODE_ENV",
              "value": "dev",
            },
          ],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div>
          <styled.span
            error={false}
          />
          <StateInjector
            __debug={false}
            watchPath={[Function]}
          >
            [Function]
          </StateInjector>
        </styled.div>,
        "_debugID": 2,
        "_renderedOutput": <styled.div>
          <styled.span
            error={false}
          />
          <StateInjector
            __debug={false}
            watchPath={[Function]}
          >
            [Function]
          </StateInjector>
        </styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <PlanConfigList
    items={
      Object {
        "schemas": Object {
          "name": Object {
            "default": "",
            "placeholder": "name",
            "type": "string",
          },
          "value": Object {
            "default": "",
            "placeholder": "value",
            "type": "string",
          },
        },
        "type": "object",
      }
    }
    onChange={[Function]}
    values={
      Array [
        Object {
          "name": "NODE_PLAN",
          "value": "web",
        },
        Object {
          "name": "NODE_ENV",
          "value": "dev",
        },
      ]
    }
  />,
}
`;
